### DISCUSSION TITLE
Getting the user ID in the job handler so as to process Stripe webhook payload

### DISCUSSION THREAD

**Discussion by hondaman900** [spatie/laravel-stripe-webhooks#148] [Q&A]
So with Laravel-stripe-webhooks I can successfully receive and parse a Stripe payload in my dev system, with the help of Stripe CLI and listening for my endpoint. However, if I try to get the current user ID in the job, so as to process the payment they just made, the `Auth::User()` object comes back null. I can't get any user parameters, nor any `dd()` responses out of the job code. I can, however, write to the logfile.

What am I missing? Shouldn't I still be able to reference `Auth::User()`, or is there another designated way to get my user's ID in the processing code?

This is my job used as the Stripe endpoint:

**CustomerCreatedJob.php**
```
<?php

namespace App\Jobs\StripeWebhooks;

use App\User;
use Auth;
use App\Payment;
use Illuminate\Bus\Queueable;
use Illuminate\Contracts\Queue\ShouldQueue;
use Illuminate\Foundation\Bus\Dispatchable;
use Illuminate\Queue\InteractsWithQueue;
use Illuminate\Queue\SerializesModels;
use Spatie\WebhookClient\Models\WebhookCall;
use Request;
use IlluminateSupportFacadesLog;
use Illuminate\Support\Facades\Log;
use Stripe\Event;

class CustomerCreatedJob implements ShouldQueue
{
    use Dispatchable, InteractsWithQueue, Queueable, SerializesModels;

     /** @var \Spatie\WebhookClient\Models\WebhookCall */
     public $webhookCall;


       public function __construct(WebhookCall $webhookCall)
    {
        $this->webhookCall = $webhookCall;
        $this->middleware('auth'); 
    }

    public function handle(WebhookCall $webhookCall)
    {
        $charge = $this->webhookCall->payload['data']['object'];
        $payload = $this->webhookCall->payload; //@file_get_contents('php://input');
        $user = Auth::User();
        
        Log::info(Auth::User());
        Log::info("Customer ID: ".$user);
        Log::info("Stripe ID: ".$charge['id']);
        Log::info("Customer Name: ".$charge['name']);
        Log::info("Customer Phone: ".$charge['phone']);
        Log::info("Customer Email: ".$charge['email']);
        Log::info("Customer Address: ");
        Log::info($charge['address']['line1']);
        Log::info($charge['address']['city'].", ".$charge['address']['state']." ".$charge['address']['postal_code']);
        //Log::info();
        Log::info($charge['address']['country']);
        Log::info("Amount: ".$charge['subscriptions']['data']['0']['items']['data']['0']['plan']['amount_decimal']);
        Log::info("Plan: ".$charge['subscriptions']['data']['0']['items']['data']['0']['plan']['nickname']);
    }
}
```

The route:
**Web.php**
`Route::stripeWebhooks('stripe-webhook');`

In **stripe-webhooks.php**:

```
    'jobs' => [
    'customer_created' => \App\Jobs\StripeWebhooks\CustomerCreatedJob::class,
    ],
```

This is my logfile output, line-for-line where you can see the first two lines show blanks for the user data.

```
[2023-01-17 12:38:25] local.INFO:   
[2023-01-17 12:38:25] local.INFO: Customer ID:   
[2023-01-17 12:38:25] local.INFO: Stripe ID: cus_NBjOmJVbTFXhdE  
[2023-01-17 12:38:25] local.INFO: Customer Name: Purple Rain  
[2023-01-17 12:38:25] local.INFO: Customer Phone: +12098874666  
[2023-01-17 12:38:25] local.INFO: Customer Email: purple@adminc.om  
[2023-01-17 12:38:25] local.INFO: Customer Address:   
[2023-01-17 12:38:25] local.INFO: 95446 Woodbridge Parkway  
[2023-01-17 12:38:25] local.INFO: Pensacola, FL 32034  
[2023-01-17 12:38:25] local.INFO: US  
[2023-01-17 12:38:25] local.INFO: Amount: 24995  
[2023-01-17 12:38:25] local.INFO: Plan: Ultimate  
```
Seems like, an essential part to processing webhooks is to know who the authenticated user is that make them, right? I must be missing something....

**Comment by gmgarrison** [comment]
The webhook is processing an API call that Stripe is making to your system; it's not a user-initiated action and it doesn't come through your web routing or authentication (which is why there's a secret to verify against).

The easiest way I've found is to add metadata in the payment intent - that gets sent back to you in the webhook payload so you can use it to see who was it that did the buying (and anything else you want to put in the metadata).

**Reply by hondaman900** [reply]
Thank you @gmgarrison. I'm using Stripe payment links for the user to purchase a sub with a free trial period. I'm not seeing anything in Stripe's dashboard to build a link incl. metadata. Am I in a bind because I need to use a Stripe payment link, or is there a way to pass parameters/metadata in the payment link? 

I'm wondering if I can pass those parameters between these disparate systems in the session variables? Or a cookie?

**Comment by hondaman900** [comment]
I got it! You CAN add parameters to the Stripe payment link that come back in the payload.

I found this article in the Stripe docs, [Use URL parameters to customize the checkout experience
](https://stripe.com/docs/payments/payment-links#url-parameters)

I can pass the `$userid` to the job through the payload and use that to find the correct user record to update. Many thanks, I think I'm good to go now  :)

**Reply by gmgarrison** [reply]
Happy to hear it! I don't think I helped too much but it's always great to share in a win. :)
